src/main/java/picard/sam/util/ReadNameParser.java

```diff
@@ -136,18 +138,19 @@ private boolean readLocationInformation(final String readName, final PhysicalLoc
     }
 
     public boolean addLocationInformation(final String readName, final PhysicalLocation loc){
-        final PhysicalLocation location = locationCache.getIfPresent(readName);
-        if (location == null) {
-            if (readLocationInformation(readName, loc)) {
-                locationCache.put(readName, loc);
+        if (readName != readNameStored) {
+            boolean b = readLocationInformation(readName, loc);
+            if (b) {
+                readNameStored = readName;
+                physicalLocationStored = loc;
```

Comment:
since you don't control `loc`, and it isn't immutable, it would be better to copy over the contents into `physicalLocationStored` rather than keeping a reference...